pennylane/ops/op_math/controlled_class.py

```diff
@@ -0,0 +1,380 @@
+                target_mat = sparse.lil_matrix(self.base.matrix())
+            except operation.MatrixUndefinedError as e:
+                raise operation.SparseMatrixUndefinedError from e
+
+        num_target_states = 2 ** len(self.target_wires)
+        num_control_states = 2 ** len(self.control_wires)
+        total_states = num_target_states * num_control_states
+
+        start_ind = self._control_int * num_target_states
+        end_ind = start_ind + num_target_states
+
+        m = sparse.eye(total_states, format="lil", dtype=target_mat.dtype)
+
+        m[start_ind:end_ind, start_ind:end_ind] = target_mat
+
+        return m.asformat(format=format)
+
+    def eigvals(self):
+        base_eigvals = self.base.eigvals()
+        num_target_wires = len(self.target_wires)
+        num_control_wires = len(self.control_wires)
+
+        total = 2 ** (num_target_wires + num_control_wires)
+        ones = np.ones(total - len(base_eigvals))
+
+        return qmlmath.concatenate([ones, base_eigvals])
+
+    def diagonalizing_gates(self):
+        return self.base.diagonalizing_gates()
+
+    def decomposition(self):
+        if not all(self.control_values):
+            d = [
+                qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val
+            ]
+            d += [Controlled(self.base, self.control_wires, work_wires=self.work_wires)]
+            d += [
+                qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val
+            ]
+
+            return d
+        # More to come.  This will be an extensive PR in and of itself.
+        return super().decomposition()
+
+    def generator(self):
+        sub_gen = self.base.generator()
+        proj_tensor = operation.Tensor(*(qml.Projector([1], wires=w) for w in self.control_wires))
+        return 1.0 * proj_tensor @ sub_gen
+
+    def adjoint(self):
+        return Controlled(
+            self.base.adjoint(), self.control_wires, self.control_values, self.work_wires
+        )
+
+    def pow(self, z):
+        base_pow = self.base.pow(z)
+        return [
+            Controlled(op, self.control_wires, self.control_values, self.work_wires)
+            for op in base_pow
+        ]
+
+
+class ControlledOp(Controlled, operation.Operation):
+    """Operation-specific methods and properties for the :class:`~.ops.op_math.Controlled` class.
+
+    When an :class:`~.operation.Operation` is provided to the :class:`~.ops.op_math.Controlled`
+    class, this type is constructed instead. It adds some additional :class:`~.operation.Operation`
+    specific methods and properties.
+
+    When we no longer rely on certain functionality through ``Operation``, we can get rid of this
+    class.
+    """
+
+    def __new__(cls, *_, **__):
+        # overrides dispatch behavior of ``Controlled``
+        return object.__new__(cls)
+
+    @property
+    def _inverse(self):
+        return False
+
+    @_inverse.setter
```

Comment:
Same here, if `_inverse` has a setter function, shouldn't it be a public property?